bin/compile

```diff
@@ -0,0 +1,181 @@
+install_rustup() {
+  topic "Preparing rustup"
+
+  if is_rustup_already_installed; then
+    subtopic "rustup already installed. Updating and ensuring correct version"
+
+    rustup self update
+    rustup update "$RUST_VERSION"
+    rustup default "$RUST_VERSION"
+  else
+    pushd "$CACHE_DIR" || exit 1
+
+    subtopic "Downloading rustup"
+    curl https://sh.rustup.rs -sSf > rustup.sh
+    chmod u+x rustup.sh
+
+    subtopic "Using rustup to install Rust channel $RUST_VERSION"
+    ./rustup.sh -y --default-toolchain "$RUST_VERSION"
+    rm rustup.sh
+
+    popd || exit 1
+  fi
+}
+
+# Run a Cargo build assuming that there is a Cargo.toml file in the root of the build directory
+# This task is installing all of the Rust binaries and dependencies so that we can then later
+# actually compile our Ruby native extensions
+cargo_build() {
+  topic "Running cargo build"
+
+  pushd "$BUILD_DIR" || exit 1
+
+  cargo build --release
+
+  popd || exit 1
+}
+
+# Now that all of our Rust dependencies are installed, we can actually run the Rake task
+# that builds our Ruby native extensions.
+#
+# NOTE: the Heroku env var that specifies the rake task could be multiple tasks, separated by spaces,
+# (e.g. "compile:my_gem clean").
+rake_build() {
+  topic "Running rake $RAKE_TASK"
+
+  pushd "$BUILD_DIR" || exit 1
+
+  # The RAKE_TASK might be multiple tasks split by spaces so
+  # we run this command with intention of re-splitting
+  # shellcheck disable=SC2086
+  bundle exec rake $RAKE_TASK
+
+  popd || exit 1
+}
+
+# The Cargo build process creates a bunch of binary artifacts in the `./target` dir of the app.
+# Once we have done the final rake compilation of our native extensions, we can remove these
+# so that they don't inflate the slug size.
+#
+# Based on some hand experimentation, deleting this directory after compilation shaved off ~70MB
+# from the final slug size (for a fairly small native extension).
+cleanup_build_target() {
+  topic "Cleaning up intermediate cargo build artifacts"
+
+  pushd "$BUILD_DIR" || exit 1
+  # Assert that $BUILD_DIR/target exists
+  if [ ! -d target ]; then
+    echo "$BUILD_DIR/target does not exist?"
+    ls -la "$BUILD_DIR"
+    exit 1
+  fi
+
+  rm -rf ./target
+
+  popd || exit 1
+}
+
+# This buildpack assumes that heroku-buildpack-apt runs before it. That buildpack installs apt packages
+# to $BUILD_DIR/.apt which by default will be included in our final slug. We don't need any of these
+# binaries at runtime now that our native extension is compiled, so we can remove it.
+#
+# Based on some hand experimentation, this dropped the final slug size by ~160MB.
+cleanup_apt_artifacts() {
+  topic "Removing .apt artifacts to minimize slug size"
+
+  pushd "$BUILD_DIR" || exit 1
+
+  # Assert that $BUILD_DIR/.apt exists
+  if [ ! -d .apt ]; then
+    echo "$BUILD_DIR/.apt does not exist?"
+    ls -la "$BUILD_DIR"
+    exit 1
+  fi
+
+  rm -rf ./.apt
```

Comment:
I'd suggest tracking the build deps separately somehow, so that this doesn't conflict so hard with other legit uses of heroku-buildpack-apt.

Reply:
I don't think we can control where the apt buildpack spits things out